sdk/testing/stepwise/stepwise.go

```diff
@@ -0,0 +1,351 @@
+// Package stepwise offers types and functions to enable black-box style tests
+// that are executed in defined set of steps. Stepwise utilizes "Environments" which
+// setup a running instance of Vault and provide a valid API client to execute
+// user defined steps against.
+package stepwise
+
+import (
+	"fmt"
+	"os"
+	"testing"
+
+	log "github.com/hashicorp/go-hclog"
+
+	"github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/sdk/helper/consts"
+	"github.com/hashicorp/vault/sdk/helper/logging"
+)
+
+// StepOperation defines operations each step could preform
+type StepOperation string
+
+const (
+	WriteOperation  StepOperation = "create"
+	UpdateOperation               = "update"
+	ReadOperation                 = "read"
+	DeleteOperation               = "delete"
+	ListOperation                 = "list"
+	HelpOperation                 = "help"
+)
+
+// StepwiseEnvironment is the interface Environments need to implement to be used in
+// Case to execute each Step
+type StepwiseEnvironment interface {
+	Setup() error
+	Client() (*api.Client, error)
```

Comment:
Can `Client()` get a comment on it - In particular I'd like it to be clear whether or not this should be creating a new client every time it's called vs returning a cached client.

Reply:
Updated in 359a4ff15, along with other methods in the interface